finality-aleph/src/sync/handler.rs

```diff
@@ -0,0 +1,486 @@
+    sync::{
+        data::{NetworkData, State},
+        forest::{Error as ForestError, Forest, JustificationAddResult},
+        BlockIdFor, BlockIdentifier, ChainStatus, Finalizer, Header, Justification, PeerId,
+        Verifier,
+    },
+};
+
+/// How many justifications we will send at most in response to a broadcast.
+const MAX_SMALL_JUSTIFICATION_BATCH: u32 = 10;
+// Silly workaround to make range matching actually work...
+const MAX_SMALL_JUSTIFICATION_BATCH_PLUS_ONE: u32 = MAX_SMALL_JUSTIFICATION_BATCH + 1;
+/// How many justifications we will send at most in response to an explicit query.
+const MAX_JUSTIFICATION_BATCH: u32 = 100;
+
+/// Handler for data incoming from the network.
+pub struct Handler<I: PeerId, J: Justification, CS: ChainStatus<J>, V: Verifier<J>, F: Finalizer<J>>
+{
+    chain_status: CS,
+    verifier: V,
+    finalizer: F,
+    forest: Forest<I, J>,
+    period: SessionPeriod,
+}
+
+/// What actions can the handler recommend as a reaction to some data.
+#[derive(Clone, Debug)]
+pub enum SyncActions<J: Justification> {
+    /// A response for the peer that sent us the data.
+    Response(NetworkData<J>),
+    /// A task that should be performed periodically. At the moment these are only requests for blocks,
+    /// so it always contains the id of the block.
+    Task(BlockIdFor<J>),
+    /// Do nothing.
+    Noop,
+}
+
+impl<J: Justification> SyncActions<J> {
+    fn noop() -> Self {
+        SyncActions::Noop
+    }
+
+    fn response(response: NetworkData<J>) -> Self {
+        SyncActions::Response(response)
+    }
+
+    fn task(id: BlockIdFor<J>) -> Self {
+        SyncActions::Task(id)
+    }
+}
+
+/// What can go wrong when handling a piece of data.
+#[derive(Clone, Debug)]
+pub enum Error<J: Justification, CS: ChainStatus<J>, V: Verifier<J>, F: Finalizer<J>> {
+    Verifier(V::Error),
+    ChainStatus(CS::Error),
+    Finalizer(F::Error),
+    Forest(ForestError),
+    NoParent,
+    MissingJustification,
+}
+
+impl<J: Justification, CS: ChainStatus<J>, V: Verifier<J>, F: Finalizer<J>> From<ForestError>
+    for Error<J, CS, V, F>
+{
+    fn from(e: ForestError) -> Self {
+        Error::Forest(e)
+    }
+}
+
+impl<I: PeerId, J: Justification, CS: ChainStatus<J>, V: Verifier<J>, F: Finalizer<J>>
+    Handler<I, J, CS, V, F>
+{
+    /// New handler with the provided chain interfaces.
+    pub fn new(
+        chain_status: CS,
+        verifier: V,
+        finalizer: F,
+        period: SessionPeriod,
+    ) -> Result<Self, Error<J, CS, V, F>> {
+        let top_finalized = chain_status
+            .top_finalized()
+            .map_err(Error::ChainStatus)?
+            .header()
+            .id();
+        let forest = Forest::new(top_finalized);
+        Ok(Handler {
+            chain_status,
+            verifier,
+            finalizer,
+            forest,
+            period,
+        })
+    }
+
+    fn large_justification_batch_from(
```

Comment:
The reasoning behind the naming is that `top_understandable_for`(someone who has the perspective of this block) and `*justification_batch_(starting_)from`. I'm not sure what `for` would stand for in your proposal, so if we wanted to unify this I would rather go for changing to `top_understandable_from`(the perspective of this block). WDYT?

Technically we could, but all the uses get a value anyway (from the `id` method of `Header`), so this wouldn't really help with anything.

Reply:
Ok, `from` and taking value sounded confusing to me as it looked like a constructor. If it stands for "starting from" it makes sense to me. Feel free to keep it as it is.

As of getting the value, now you have to pass/clone it, and after that it clones inside again. So potentially 2 clones instead of one. This is something that I thought of lately when looking at sync code, that in a lot of places we take value, where instead we could just take reference.